pkg/controller/clusters/status.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	return nil, nil
+}
+
+// UpdateControlPlaneInitializedCondition updates the ControlPlaneInitialized condition if it hasn't already been set.
+// This condition should be set only once.
+func UpdateControlPlaneInitializedCondition(cluster *anywherev1.Cluster, kcp *controlplanev1.KubeadmControlPlane) error {
+	// Return early if the ControlPlaneInitializedCondition is already "True"
+	if conditions.IsTrue(cluster, anywherev1.ControlPlaneInitializedCondition) {
+		return nil
+	}
+
+	if kcp == nil {
+		conditions.Set(cluster, waitingForCPInitializedCondition())
+		return nil
+	}
+
+	con, err := checkKubeadmControlPlaneError(cluster, anywherev1.ControlPlaneInitializedCondition, kcp, controlplanev1.AvailableCondition)
+	if err != nil {
+		return err
+	}
+
+	if con != nil {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneInitializedCondition, con.Reason, con.Severity, con.Message)
+		return nil
+	}
+
+	available := conditions.IsTrue(kcp, controlplanev1.AvailableCondition)
+	if !available {
+		conditions.Set(cluster, waitingForCPInitializedCondition())
+		return nil
+	}
+
+	conditions.MarkTrue(cluster, anywherev1.ControlPlaneInitializedCondition)
+	return nil
+}
+
+func waitingForCPInitializedCondition() *anywherev1.Condition {
+	return conditions.FalseCondition(anywherev1.ControlPlaneInitializedCondition, anywherev1.WaitingForControlPlaneInitializedReason, clusterv1.ConditionSeverityInfo, "The first control plane instance is not available yet")
+}
+
+// UpdateControlPlaneReadyCondition updates the ControlPlaneReady condition. It checks whether the KubeadmControlPlane is ready
+// and number of ready vs expected replicas are accounted for.
+func UpdateControlPlaneReadyCondition(cluster *anywherev1.Cluster, kcp *controlplanev1.KubeadmControlPlane) error {
+	initializedCondition := conditions.Get(cluster, anywherev1.ControlPlaneInitializedCondition)
+	if initializedCondition.Status != "True" {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, initializedCondition.Reason, initializedCondition.Severity, initializedCondition.Message)
+		return nil
+	}
+
+	if kcp == nil {
+		return fmt.Errorf("expected reference to kubeadmcontrolplane, but got nil instead ")
+	}
+
+	statusUpToDate := kcp.Status.ObservedGeneration == kcp.ObjectMeta.Generation
+	// We make sure to check that the status is up to date before using it
+	if !statusUpToDate {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, anywherev1.PendingUpdateReason, clusterv1.ConditionSeverityInfo, "")
+		return nil
+	}
+
+	con, err := checkKubeadmControlPlaneError(cluster, anywherev1.ControlPlaneReadyCondition, kcp, clusterv1.ReadyCondition)
+	if err != nil {
+		return err
+	}
+
+	if con != nil {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, con.Reason, con.Severity, con.Message)
+		return nil
+	}
+
+	// The control plane should be marked ready when the total number of nodes specified in the spec is
```

Comment:
Do we have to do this math ourselves? Does cluster-api not keep track of this already?